File: vault-ios/VaultCommon/Features/Repo/RepoInfoScreen.swift
import SwiftUI
import VaultMobile

public class RepoInfoScreenViewModel: ObservableObject {
    public let container: Container
    public let navController: MainNavController
    public let repoId: String

    @Published public var biometricUnlockEnabled = false

    public init(container: Container, navController: MainNavController, repoId: String) {
        self.container = container
        self.navController = navController
        self.repoId = repoId

        self._biometricUnlockEnabled = Published(initialValue: checkBiometricUnlockEnabled())
    }

    public func updateBiometricUnlockEnabled() {
        self.biometricUnlockEnabled = checkBiometricUnlockEnabled()
    }

    public func checkBiometricUnlockEnabled() -> Bool {
        do {
            return try container.keychainRepoPasswordStorage.hasPassword(repoId: repoId)
        } catch {
            print("RepoInfoScreen keychainRepoPasswordStorage.hasPassword error: \(error)")

            return false
        }
    }

    public func disableBiometricUnlock() {
        do {
            try container.keychainRepoPasswordStorage.removePassword(repoId: repoId)

            updateBiometricUnlockEnabled()
        } catch {
            print("RepoInfoScreen keychainRepoPasswordStorage.removePassword error: \(error)")
        }
    }
}

public struct RepoInfoScreen: View {
    @ObservedObject var vm: RepoInfoScreenViewModel

    @ObservedObject private var info: Subscription<RepoInfo>

    @State private var autoLockAfterOptionsPresented = false

    public init(vm: RepoInfoScreenViewModel) {
        self.vm = vm

        self.info = Subscription(
            mobileVault: vm.container.mobileVault,
            subscribe: { v, cb in
                v.reposRepoSubscribe(repoId: vm.repoId, cb: cb)
            },
            getData: { v, id in
                v.reposRepoData(id: id)
            })
    }

    public var body: some View {
        if let repo = info.data?.repo {
            let unlocked = Binding(
                get: { repo.state == .unlocked },
                set: { value in
                    if value {
                        vm.container.sheets.show(
                            name: "repoInfoUnlock",
                            viewModel: RepoUnlockScreenViewModel(
                                container: vm.container, repoId: repo.id,
                                setupBiometricUnlockVisible: false)
                        ) { vm, hide in
                            RepoUnlockSheet(vm: vm, onDismiss: hide)
                        }
                    } else {
                        vm.container.mobileVault.reposLockRepo(repoId: vm.repoId)
                    }
                })

            let biometricUnlockEnabled = Binding(
                get: {
                    vm.biometricUnlockEnabled
                },
                set: { value in
                    if value {
                        vm.container.sheets.show(
                            viewModel: RepoSetupBiometricUnlockSheetViewModel(
                                container: vm.container, repoId: vm.repoId),
                            onHide: {
                                vm.updateBiometricUnlockEnabled()
                            },
                            content: { setupVm, hide in
                                RepoSetupBiometricUnlockSheet(vm: setupVm, onDismiss: hide)
                            })
                    } else {
                        vm.disableBiometricUnlock()
                    }
                })

            let repoAutoLockOnAppHidden = Binding(
                get: {
                    repo.autoLock.onAppHidden
                },
                set: { value in
                    vm.container.mobileVault.reposSetAutoLock(
                        repoId: vm.repoId,
                        autoLock: RepoAutoLock(after: repo.autoLock.after, onAppHidden: value))
                })

            List {
                Section {
                    HStack {
                        Toggle(isOn: unlocked) {
                            VStack(alignment: .leading) {
                                Text(
                                    repo.state == .unlocked
                                        ? LocalizedStringResource(
                                            "ios.repo_info.unlocked.label",
                                            defaultValue: "Unlocked",
                                            bundle: #bundle,
                                            comment:
                                                "Toggle label in Safe Box settings when the Safe Box is currently unlocked."
                                        )
                                        : LocalizedStringResource(
                                            "ios.repo_info.locked.label",
                                            defaultValue: "Locked",
                                            bundle: #bundle,
                                            comment:
                                                "Toggle label in Safe Box settings when the Safe Box is currently locked."
                                        )
                                )
                                .padding(.bottom, 0.5)
                                Text(
                                    LocalizedStringResource(
                                        "ios.repo_info.unlock.description",
                                        defaultValue: "Unlock or lock the Safe Box",
                                        bundle: #bundle,
                                        comment:
                                            "Description under the lock state toggle in Safe Box settings."
                                    )
                                )
                                .font(.system(.footnote))
                                .foregroundColor(Color(.secondaryLabel))
                            }
                        }
                    }
                    .padding(.vertical, 2)

                    HStack {
                        Toggle(isOn: biometricUnlockEnabled) {
                            VStack(alignment: .leading) {
                                Text(
                                    LocalizedStringResource(
                                        "ios.repo_info.biometric_unlock.label",
                                        defaultValue: "Biometric unlock",
                                        bundle: #bundle,
                                        comment:
                                            "Toggle label for enabling biometric unlock in Safe Box settings."
                                    )
                                )
                                .padding(.bottom, 0.5)
                                Text(
                                    LocalizedStringResource(
                                        "ios.repo_info.biometric_unlock.description",
                                        defaultValue: "Use biometrics to unlock the Safe Box",
                                        bundle: #bundle,
                                        comment:
                                            "Description under the biometric unlock toggle in Safe Box settings."
                                    )
                                )
                                .font(.system(.footnote))
                                .foregroundColor(Color(.secondaryLabel))
                            }
                        }
                    }
                    .padding(.vertical, 2)

                    Button {
                        autoLockAfterOptionsPresented = true
                    } label: {
                        VStack(alignment: .leading) {
                            Text(
                                LocalizedStringResource(
                                    "ios.repo_info.auto_lock_after.label",
                                    defaultValue: "Automatically lock after",
                                    bundle: #bundle,
                                    comment:
                                        "Label for the row that opens auto-lock timeout options in Safe Box settings."
                                )
                            )
                            .padding(.bottom, 0.5)
                            .foregroundColor(Color(.label))

                            Text(repoAutoLockAfterDisplay(repo.autoLock.after))
                                .font(.system(.footnote))
                                .foregroundColor(Color(.secondaryLabel))
                        }
                    }
                    .padding(.vertical, 2)
                    .confirmationDialog(
                        Text(
                            LocalizedStringResource(
                                "ios.repo_info.auto_lock_after.label",
                                defaultValue: "Automatically lock after",
                                bundle: #bundle,
                                comment:
                                    "Title of the confirmation dialog for choosing auto-lock timeout."
                            )
                        ),
                        isPresented: $autoLockAfterOptionsPresented
                    ) {
                        ForEach(
                            getRepoAutoLockAfterOptions(current: repo.autoLock.after), id: \.self
                        ) { option in
                            Button {
                                vm.container.mobileVault.reposSetAutoLock(
                                    repoId: vm.repoId,
                                    autoLock: RepoAutoLock(
                                        after: option, onAppHidden: repo.autoLock.onAppHidden))
                            } label: {
                                Text(repoAutoLockAfterDisplay(option))
                            }
                        }
                    }

                    HStack {
                        Toggle(isOn: repoAutoLockOnAppHidden) {
                            VStack(alignment: .leading) {
                                Text(
                                    LocalizedStringResource(
                                        "ios.repo_info.lock_when_app_hidden.label",
                                        defaultValue: "Lock when app hidden",
                                        bundle: #bundle,
                                        comment:
                                            "Toggle label for locking the Safe Box when app goes to background."
                                    )
                                )
                                .padding(.bottom, 0.5)
                                Text(
                                    LocalizedStringResource(
                                        "ios.repo_info.lock_when_app_hidden.description",
                                        defaultValue: "When switching apps or locking the screen",
                                        bundle: #bundle,
                                        comment:
                                            "Description under the lock-on-background toggle in Safe Box settings."
                                    )
                                )
                                .font(.system(.footnote))
                                .foregroundColor(Color(.secondaryLabel))
                            }
                        }
                    }
                    .padding(.vertical, 2)
                }

                Section {
                    HStack {
                        Button {
                            vm.navController.push(.repoRemove(repoId: vm.repoId))
                        } label: {
                            VStack(alignment: .leading) {
                                Text(
                                    LocalizedStringResource(
                                        "ios.repo_info.destroy_repo.label",
                                        defaultValue: "Destroy Safe Box…",
                                        bundle: #bundle,
                                        comment:
                                            "Row label in Safe Box settings that opens the destroy Safe Box dialog."
                                    )
                                )
                                .padding(.bottom, 0.5)
                                .foregroundColor(Color(.label))
                                Text(
                                    LocalizedStringResource(
                                        "ios.repo_info.destroy_repo.description",
                                        defaultValue: "Verify Safe Key and destroy the Safe box",
                                        bundle: #bundle,
                                        comment:
                                            "Description under the destroy Safe Box row in Safe Box settings."
                                    )
                                )
                                .font(.system(.footnote))
                                .foregroundColor(Color(.secondaryLabel))
                            }
                        }
                    }
                    .padding(.vertical, 2)
                }
            }
            .navigationTitle(repo.name)
        }
    }

    func repoAutoLockAfterDisplay(_ after: RepoAutoLockAfter) -> LocalizedStringResource {
        switch after {
        case .noLimit:
            return LocalizedStringResource(
                "ios.repo_info.auto_lock_after.no_time_limit",
                defaultValue: "No time limit",
                bundle: #bundle,
                comment:
                    "Auto-lock timeout option label meaning the safe box should not auto-lock by time."
            )
        case .inactive1Minute:
            return LocalizedStringResource(
                "ios.repo_info.auto_lock_after.inactive_1_minute",
                defaultValue: "1 minute of inactivity",
                bundle: #bundle,
                comment: "Auto-lock timeout option label for one minute of inactivity."
            )
        case .inactive5Mininutes:
            return LocalizedStringResource(
                "ios.repo_info.auto_lock_after.inactive_5_minutes",
                defaultValue: "5 minutes of inactivity",
                bundle: #bundle,
                comment: "Auto-lock timeout option label for five minutes of inactivity."
            )
        case .inactive10Minutes:
            return LocalizedStringResource(
                "ios.repo_info.auto_lock_after.inactive_10_minutes",
                defaultValue: "10 minutes of inactivity",
                bundle: #bundle,
                comment: "Auto-lock timeout option label for ten minutes of inactivity."
            )
        case .inactive30Minutes:
            return LocalizedStringResource(
                "ios.repo_info.auto_lock_after.inactive_30_minutes",
                defaultValue: "30 minutes of inactivity",
                bundle: #bundle,
                comment: "Auto-lock timeout option label for thirty minutes of inactivity."
            )
        case .inactive1Hour:
            return LocalizedStringResource(
                "ios.repo_info.auto_lock_after.inactive_1_hour",
                defaultValue: "1 hour of inactivity",
                bundle: #bundle,
                comment: "Auto-lock timeout option label for one hour of inactivity."
            )
        case .inactive2Hours:
            return LocalizedStringResource(
                "ios.repo_info.auto_lock_after.inactive_2_hours",
                defaultValue: "2 hours of inactivity",
                bundle: #bundle,
                comment: "Auto-lock timeout option label for two hours of inactivity."
            )
        case .inactive4Hours:
            return LocalizedStringResource(
                "ios.repo_info.auto_lock_after.inactive_4_hours",
                defaultValue: "4 hours of inactivity",
                bundle: #bundle,
                comment: "Auto-lock timeout option label for four hours of inactivity."
            )
        case .custom(let seconds):
            return LocalizedStringResource(
                "ios.repo_info.auto_lock_after.custom",
                defaultValue: "Custom (\(seconds) seconds)",
                bundle: #bundle,
                comment: "Auto-lock timeout option label for a custom number of inactive seconds."
            )
        }
    }

    func getRepoAutoLockAfterOptions(current: RepoAutoLockAfter) -> [RepoAutoLockAfter] {
        var options: [RepoAutoLockAfter] = [
            .noLimit,
            .inactive1Minute,
            .inactive5Mininutes,
            .inactive10Minutes,
            .inactive30Minutes,
            .inactive1Hour,
            .inactive2Hours,
            .inactive4Hours,
        ]

        switch current {
        case .custom(_): options.append(current)
        default: ()
        }

        return options
    }
}

File: vault-android/app/src/main/java/net/koofr/vault/features/repo/RepoInfoScreen.kt
package net.koofr.vault.features.repo

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.selection.selectable
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelStore
import dagger.hilt.android.lifecycle.HiltViewModel
import net.koofr.vault.LocalSnackbarHostState
import net.koofr.vault.MobileVault
import net.koofr.vault.RepoAutoLockAfter
import net.koofr.vault.RepoState
import net.koofr.vault.features.mobilevault.AndroidSecureStorage
import net.koofr.vault.features.mobilevault.subscribe
import net.koofr.vault.features.navigation.LocalNavController
import net.koofr.vault.features.repounlock.RepoUnlockDialog
import net.koofr.vault.utils.WithCustomViewModelStore
import javax.inject.Inject

@HiltViewModel
class RepoInfoScreenViewModel @Inject constructor(
    val mobileVault: MobileVault,
    androidSecureStorage: AndroidSecureStorage,
    savedStateHandle: SavedStateHandle,
) : ViewModel() {
    val repoId: String = savedStateHandle.get<String>("repoId")!!

    private val biometricsHelper: RepoPasswordBiometricsHelper =
        RepoPasswordBiometricsHelper(repoId, androidSecureStorage)

    val biometricUnlockEnabled = mutableStateOf(biometricsHelper.isBiometricUnlockEnabled())

    val repoUnlockDialog = mutableStateOf<ViewModelStore?>(null)
    val setupBiometricUnlockVisible = mutableStateOf(false)
    val repoAutoLockAfterDialogVisible = mutableStateOf(false)

    fun updateBiometricUnlockEnabled() {
        biometricUnlockEnabled.value = biometricsHelper.isBiometricUnlockEnabled()
    }

    fun removeBiometricUnlock() {
        biometricsHelper.removeBiometricUnlock()

        updateBiometricUnlockEnabled()
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RepoInfoScreen(
    vm: RepoInfoScreenViewModel = hiltViewModel(),
) {
    val navController = LocalNavController.current

    val repo = subscribe(
        { v, cb -> v.reposRepoSubscribe(repoId = vm.repoId, cb = cb) },
        { v, id -> v.reposRepoData(id = id) },
    )

    LaunchedEffect(Unit) {
        // if user clicks unlock and the biometric unlocks fails we need to
        // recheck if it is enabled
        vm.updateBiometricUnlockEnabled()
    }

    Scaffold(topBar = {
        TopAppBar(title = {
            Text(
                repo.value?.repo?.name ?: "",
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        })
    }, snackbarHost = { SnackbarHost(LocalSnackbarHostState.current) }) { paddingValues ->
        repo.value?.repo?.let { repo ->
            Column(modifier = Modifier.padding(paddingValues)) {
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(15.dp),
                ) {
                    Column(
                        modifier = Modifier
                            .padding(0.dp, 0.dp, 10.dp, 0.dp)
                            .weight(1.0f),
                    ) {
                        Text(
                            text = repo.state.let {
                                when (it) {
                                    RepoState.LOCKED -> "Locked"
                                    RepoState.UNLOCKED -> "Unlocked"
                                }
                            },
                            style = MaterialTheme.typography.bodyLarge,
                            modifier = Modifier
                                .padding(0.dp, 0.dp, 0.dp, 2.dp),
                        )
                        Text(
                            text = "Unlock or lock the Safe Box",
                            style = MaterialTheme.typography.bodyMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                    Switch(
                        checked = repo.state.let {
                            when (it) {
                                RepoState.LOCKED -> false
                                RepoState.UNLOCKED -> true
                            }
                        },
                        onCheckedChange = {
                            if (it) {
                                vm.repoUnlockDialog.value = ViewModelStore()
                            } else {
                                vm.mobileVault.reposLockRepo(repoId = vm.repoId)
                            }
                        },
                        modifier = Modifier.semantics {
                            contentDescription = repo.state.let {
                                when (it) {
                                    RepoState.LOCKED -> "Locked"
                                    RepoState.UNLOCKED -> "Unlocked"
                                }
                            }
                        },
                    )
                }

                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(15.dp),
                ) {
                    Column(
                        modifier = Modifier
                            .padding(0.dp, 0.dp, 10.dp, 0.dp)
                            .weight(1.0f),
                    ) {
                        Text(
                            text = "Biometric unlock",
                            style = MaterialTheme.typography.bodyLarge,
                            modifier = Modifier
                                .padding(0.dp, 0.dp, 0.dp, 2.dp),
                        )
                        Text(
                            text = "Use biometrics to unlock the Safe Box",
                            style = MaterialTheme.typography.bodyMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                    Switch(
                        checked = vm.biometricUnlockEnabled.value,
                        onCheckedChange = {
                            if (vm.biometricUnlockEnabled.value) {
                                vm.removeBiometricUnlock()
                            } else {
                                vm.setupBiometricUnlockVisible.value = true
                            }
                        },
                        modifier = Modifier.semantics {
                            contentDescription = "Biometric unlock"
                        },
                    )
                }

                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable(onClick = {
                            vm.repoAutoLockAfterDialogVisible.value = true
                        }),
                ) {
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(15.dp),
                    ) {
                        Column(
                            modifier = Modifier
                                .padding(0.dp, 0.dp, 10.dp, 0.dp)
                                .weight(1.0f),
                        ) {
                            Text(
                                text = "Automatically lock after",
                                style = MaterialTheme.typography.bodyLarge,
                                modifier = Modifier
                                    .padding(0.dp, 0.dp, 0.dp, 2.dp),
                            )
                            Text(
                                text = repoAutoLockAfterDisplay(repo.autoLock.after),
                                style = MaterialTheme.typography.bodyMedium,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                        }
                    }
                }

                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(15.dp),
                ) {
                    Column(
                        modifier = Modifier
                            .padding(0.dp, 0.dp, 10.dp, 0.dp)
                            .weight(1.0f),
                    ) {
                        Text(
                            text = "Lock when app hidden",
                            style = MaterialTheme.typography.bodyLarge,
                            modifier = Modifier
                                .padding(0.dp, 0.dp, 0.dp, 2.dp),
                        )
                        Text(
                            text = "When switching apps or locking the screen",
                            style = MaterialTheme.typography.bodyMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                    Switch(
                        checked = repo.autoLock.onAppHidden,
                        onCheckedChange = {
                            vm.mobileVault.reposSetAutoLock(
                                repo.id,
                                repo.autoLock.copy(onAppHidden = it),
                            )
                        },
                        modifier = Modifier.semantics {
                            contentDescription = "Biometric unlock"
                        },
                    )
                }

                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable(onClick = {
                            navController.navigate("repos/${repo.id}/remove")
                        }),
                ) {
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(15.dp),
                    ) {
                        Column(
                            modifier = Modifier
                                .padding(0.dp, 0.dp, 10.dp, 0.dp)
                                .weight(1.0f),
                        ) {
                            Text(
                                text = "Destroy Safe Box…",
                                style = MaterialTheme.typography.bodyLarge,
                                modifier = Modifier
                                    .padding(0.dp, 0.dp, 0.dp, 2.dp),
                            )
                            Text(
                                text = "Verify Safe Key and destroy the Safe Box",
                                style = MaterialTheme.typography.bodyMedium,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                        }
                    }
                }
            }

            if (vm.repoAutoLockAfterDialogVisible.value) {
                RepoAutoLockAfterDialog(
                    options = getRepoAutoLockAfterOptions(repo.autoLock.after),
                    current = repo.autoLock.after,
                    onDismiss = {
                        vm.repoAutoLockAfterDialogVisible.value = false
                    },
                    onConfirm = {
                        vm.mobileVault.reposSetAutoLock(repo.id, repo.autoLock.copy(after = it))

                        vm.repoAutoLockAfterDialogVisible.value = false
                    },
                )
            }
        }

        if (vm.setupBiometricUnlockVisible.value) {
            RepoSetupBiometricUnlockDialog(onDismiss = {
                vm.setupBiometricUnlockVisible.value = false

                vm.updateBiometricUnlockEnabled()
            })
        }

        vm.repoUnlockDialog.value?.let { store ->
            WithCustomViewModelStore(store) {
                RepoUnlockDialog(onDismiss = {
                    store.clear()

                    vm.repoUnlockDialog.value = null
                })
            }
        }
    }
}

@Composable
fun RepoAutoLockAfterDialog(
    options: List<RepoAutoLockAfter>,
    current: RepoAutoLockAfter,
    onDismiss: () -> Unit,
    onConfirm: (RepoAutoLockAfter) -> Unit,
) {
    val selected = remember { mutableStateOf(current) }

    AlertDialog(
        onDismissRequest = onDismiss,
        confirmButton = {
            TextButton(
                onClick = {
                    onConfirm(selected.value)
                },
            ) {
                Text("OK")
            }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) {
                Text("CANCEL")
            }
        },
        title = { Text("Automatically lock after") },
        text = {
            Column(
                modifier = Modifier
                    .fillMaxWidth(),
            ) {
                LazyColumn {
                    items(options) { option ->
                        Row(
                            modifier = Modifier
                                .fillMaxWidth()
                                .selectable(
                                    selected = option == selected.value,
                                    onClick = {
                                        selected.value = option
                                    },
                                ),
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            RadioButton(
                                selected = option == selected.value,
                                onClick = { selected.value = option },
                            )
                            Spacer(modifier = Modifier.width(8.dp))
                            Text(text = repoAutoLockAfterDisplay(option))
                        }
                    }
                }
            }
        },
    )
}

private fun repoAutoLockAfterDisplay(after: RepoAutoLockAfter): String {
    return when (after) {
        is RepoAutoLockAfter.NoLimit -> "No time limit"
        is RepoAutoLockAfter.Inactive1Minute -> "1 minute of inactivity"
        is RepoAutoLockAfter.Inactive5Mininutes -> "5 minutes of inactivity"
        is RepoAutoLockAfter.Inactive10Minutes -> "10 minutes of inactivity"
        is RepoAutoLockAfter.Inactive30Minutes -> "30 minutes of inactivity"
        is RepoAutoLockAfter.Inactive1Hour -> "1 hour of inactivity"
        is RepoAutoLockAfter.Inactive2Hours -> "2 hours of inactivity"
        is RepoAutoLockAfter.Inactive4Hours -> "4 hours of inactivity"
        is RepoAutoLockAfter.Custom -> "Custom (${after.seconds} seconds)"
    }
}

private fun getRepoAutoLockAfterOptions(current: RepoAutoLockAfter): List<RepoAutoLockAfter> {
    val options = mutableListOf(
        RepoAutoLockAfter.NoLimit,
        RepoAutoLockAfter.Inactive1Minute,
        RepoAutoLockAfter.Inactive5Mininutes,
        RepoAutoLockAfter.Inactive10Minutes,
        RepoAutoLockAfter.Inactive30Minutes,
        RepoAutoLockAfter.Inactive1Hour,
        RepoAutoLockAfter.Inactive2Hours,
        RepoAutoLockAfter.Inactive4Hours,
    )

    when (current) {
        is RepoAutoLockAfter.Custom -> {
            options.add(current)
        }

        else -> {}
    }

    return options
}
